flocker/control/test/test_httpapi.py

```diff
@@ -45,25 +72,344 @@ def assertGoodResult(self, method, path, expected_good_result):
+        return self.assertGoodResult(
+            b"GET", b"/version", None, {u'flocker': __version__}
+        )
 
 
 def _build_app(test):
     test.initialize()
     return DatasetAPIUserV1(test.persistence_service).app
-RealTestsAPI, MemoryTestsAPI = buildIntegrationTests(
-    APITestsMixin, "API", _build_app)
+
+RealTestsVersion, MemoryTestsVersion = buildIntegrationTests(
+    VersionTestsMixin, "Version", _build_app)
+
+
+class CreateDatasetTestsMixin(APITestsMixin):
+    """
+    Tests for the dataset creation endpoint at ``/datasets``.
+    """
+    # These addresses taken from RFC 5737 (TEST-NET-1)
+    NODE_A = b"192.0.2.1"
+    NODE_B = b"192.0.2.2"
+
+    def test_wrong_schema(self):
+        """
+        If a ``POST`` request made to the endpoint includes a body which
+        doesn't match the ``definitions/datasets`` schema, the response is an
+        error indication a validation failure.
+        """
+        return self.assertBadResult(
+            b"POST", b"/datasets",
+            {u"primary": self.NODE_A, u"junk": u"garbage"},
+            BAD_REQUEST, {
+                u'description':
+                    u"The provided JSON doesn't match the required schema.",
+                u'errors': [
+                    u"Additional properties are not allowed "
+                    u"(u'junk' was unexpected)"
+                ]
+            }
+        )
+
+    def _dataset_id_collision_test(self, primary):
+        """
+        Assert that an attempt to create a dataset with a dataset_id that is
+        already assigned somewhere on the cluster results in an error response
+        and no configuration change.
+
+        A configuration with two nodes, ``NODE_A`` and ``NODE_B``, is created.
+        ``NODE_A`` is given one unattached manifestation.  An attempt is made
+        to configure a new dataset is on ``primary`` which should be either
+        ``NODE_A`` or ``NODE_B``.
+
+        :return: A ``Deferred`` that fires with the result of the test.
+        """
+        dataset_id = unicode(uuid4())
+        existing_dataset = Dataset(dataset_id=dataset_id)
+        existing_manifestation = Manifestation(
+            dataset=existing_dataset, primary=True)
+
+        saving = self.persistence_service.save(Deployment(
+            nodes={
+                Node(
+                    hostname=self.NODE_A,
+                    other_manifestations=frozenset({existing_manifestation})
+                ),
+                Node(hostname=self.NODE_B),
+            }
+        ))
+
+        def saved(ignored):
+            return self.assertBadResult(
+                b"POST", b"/datasets",
+                {u"primary": primary, u"dataset_id": dataset_id},
+                CONFLICT,
+                {u"description": u"The provided dataset_id is already in use."}
+            )
+        posting = saving.addCallback(saved)
+
+        def failed(reason):
+            deployment = self.persistence_service.get()
```

Comment:
I wrote it this way because I was afraid of one or the other object being mutated by another part of the code.  This doesn't happen now, as far as I know, but who knows what the future might bring.  When we get to a place where `Deployment` is wholly immutable I think I'd be happy to see the test written the way you suggest.